File: server/express/src/modules/auth/auth.service.ts
import { UserRepositoryInterface, userRepository } from "../user/user.repository";
import { BcryptUtil } from "../../utils/bcrypt.util";
import { JwtUtil } from "../../utils/jwt.util";
import { logger } from "../../config/logger";
import { ApiError } from "../../internal/api.error";

export class AuthService {

  constructor(
    private userRepository: UserRepositoryInterface
  ) {}

  async registerUser(username: string, name : string, email: string, password: string){
    logger.debug(`Registering user with email ${email} and name ${name}`);
    
    const hashedPassword = await BcryptUtil.hash(password);
    const user = await this.userRepository.create(username, name, email, hashedPassword);
    return user;
  }

  async loginUser(email: string, password: string){
    logger.debug(`Logging in user with email ${email}`);
    
    const user = await this.userRepository.findByEmail(email);
    if(!user){
      throw new ApiError(404, "User not found");
    }

    const isPasswordValid = await BcryptUtil.compare(password, user.password);
    if(!isPasswordValid){
      throw new ApiError(400, "Invalid password or compare");
    }

    logger.debug("Password and email are valid")

    const accesToken =  JwtUtil.generateAccessToken(user.id, user.email);
    const refreshToken = JwtUtil.generateRefreshToken(user.id, user.email);

    const saveToken = await this.userRepository.updateRefreshToken(user.id, refreshToken);
    if(!saveToken){
      throw new Error("Error saving refresh token");
    }
    return {accesToken, refreshToken};
  }

  async refreshTokenUser(refreshToken: string){

    // Check Token
    const userToken = await this.userRepository.findToken(refreshToken);
    if(!userToken){
      throw new ApiError(402,"Invalid refresh token");
    }

    // Verify Token
    const verify = JwtUtil.verifyRefreshToken(refreshToken);
    if(!verify){
      throw new ApiError(401,"Invalid refresh token");
    }

    const accessToken = JwtUtil.generateAccessToken(userToken.id, userToken.email);
    return accessToken;
  }

  async logoutUser(refreshToken: string) {
    
    // Checktoken 
    const userToken = await this.userRepository.findToken(refreshToken);
    if (!userToken) {
      throw new ApiError(402,"Invalid refresh token");
    }

    // verify token
    const verify = JwtUtil.verifyRefreshToken(refreshToken);
    if (!verify) {
      throw new ApiError(401,"Invalid refresh token");
    }

    // delete token
    const deleteToken = await this.userRepository.deleteRefreshToken(userToken.id);
    if (!deleteToken) {
      throw new ApiError(402,"Something error to delete refresh token");
    }

    return;
  }
}

const authService = new AuthService(userRepository);
export default authService;

File: server/gofiber/internal/modules/auth/auth_service.go
package auth

import (
	ur "gofiber/internal/modules/user"
	u "gofiber/pkg/utils"
)

type AuthServiceInterface interface {
	RegisterUser(username string, name string, email string, password string) error
	LoginUser(email string) (*ur.User, error)
}

type AuthService struct {
	userRepository ur.UserRepositoryInterface
}

func NewAuthService(userRepository ur.UserRepositoryInterface) *AuthService {
	return &AuthService{
		userRepository: userRepository,
	}
}

func (s *AuthService) RegisterUser(
	username string, name string, email string, password string,
) error {

	user, err := s.userRepository.FindByEmail(email)
	if err != nil {
		return err
	}
	if user != nil {
		return u.NewAppError(409, "Email already registered")
	}

	// Hashed Password
	passwordHashed, err := u.Hashed(password)
	if err != nil {
		return u.NewAppError(500, "Failed to hash password")
	}

	if err := s.userRepository.Create(username, name, email, passwordHashed); err != nil {
		return u.NewAppError(500, "Failed to register user")
	}

	return nil
}

func (s *AuthService) LoginUser(email string, password string) (*ur.User, error) {
	user, err := s.userRepository.FindByEmail(email)
	if err != nil {
		return nil, u.NewAppError(404, "User not found")
	}
	passwordCompare, err := u.Compare(password, user.Password)
	if err != nil || !passwordCompare {
		return nil, u.NewAppError(401, "Invalid credentials")
	}

	return user, nil
}
